pkg/controller/operators/catalogtemplate/operator.go

```
@@ -0,0 +1,235 @@
+
+	// Wire CatalogSources
+	catsrcInformer := crInformerFactory.Operators().V1alpha1().CatalogSources()
+	op.lister.OperatorsV1alpha1().RegisterCatalogSourceLister(metav1.NamespaceAll, catsrcInformer.Lister())
+	catalogTemplateSrcQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "catalogSourceTemplate")
+	op.catalogSourceTemplateQueueSet.Set(metav1.NamespaceAll, catalogTemplateSrcQueue)
+	catsrcQueueInformer, err := queueinformer.NewQueueInformer(
+		ctx,
+		queueinformer.WithLogger(op.logger),
+		queueinformer.WithQueue(catalogTemplateSrcQueue),
+		queueinformer.WithInformer(catsrcInformer.Informer()),
+		queueinformer.WithSyncer(queueinformer.LegacySyncHandler(op.syncCatalogSources).ToSyncer()),
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := op.RegisterQueueInformer(catsrcQueueInformer); err != nil {
+		return nil, err
+	}
+
+	return op, nil
+}
+
+func (o *Operator) syncCatalogSources(obj interface{}) error {
+	// this is an opportunity to update the server version (regardless of any other actions for processing a catalog source)
+	o.updateServerVersion()
+
+	inputCatalogSource, ok := obj.(*v1alpha1.CatalogSource)
+	if !ok {
+		o.logger.Debugf("wrong type: %#v", obj)
+		return nil
+	}
+
+	outputCatalogSource := inputCatalogSource.DeepCopy()
+
+	logger := o.logger.WithFields(logrus.Fields{
+		"catSrcName": outputCatalogSource.GetName(),
+		"id":         queueinformer.NewLoopID(),
+	})
+	logger.Info("syncing catalog source for annotation templates")
+
+	catalogImageTemplate := catalogsource.GetCatalogTemplateAnnotation(outputCatalogSource)
+	if catalogImageTemplate == "" {
+		logger.Debug("this catalog source is not participating in template replacement")
+		// make sure the conditions are removed
+		catalogsource.RemoveStatusConditions(logger, o.client, outputCatalogSource, StatusTypeTemplatesHaveResolved, StatusTypeResolvedImage)
```

Shouldn't this handle and return the error from this function?

---

I suppose we can requeue these... same issue exists for other status functions as well

---

We don't necessarily have to handle retry, but that function is currently swallowing the error

---

Just to be clear, I am making the change to return the error, which as I understand it, will requeue the catalog source.

---

Sure, because otherwise we are failing to update. The controller itself should retry with a default backoff on failures. I'm just saying we don't necessarily need to write explicit retry logic wrapping the client call (so that the reconcile function doesn't error out) -- I don't think we want to have code that just swallows kube client errors. otherwise why are we trying to do an update?

---

makes sense. I've fix the issue and pushed.... also made some minor adjustments to logging.